### ISSUE TITLE
ETS5 doesn't use same source udp port for following requests as it has specified in hpai

### ISSUE THREAD

**Issue by elupus** [knxd/knxd#167]
ETS5 seems to switch up the source udp port it uses for new request following a CONNECT_REQUEST.  Leading to "Invalid control address" errors. 

My suspicion is that the fix is just to use the HPAI in the requests to compare with stored channel HPAI. Unsure to what port the reply should be sent. But i suspect respecting the HPAI is correct there too.

Ie on this line (and all similar)
https://github.com/knxd/knxd/blob/master/src/libserver/eibnetserver.cpp#L628

Change from:
if (compareIPAddress (p1->src, state[i]->caddr))
to
if (compareIPAddress (r1.caddr, state[i]->caddr))

It does seem to use the data port as the source, so one alternative is to compare to both caddr and daddr.

```
448	25.461187000	192.168.4.227	192.168.3.117	UDP	68	Source port: 59102  Destination port: 3671
449	25.473245000	192.168.3.117	192.168.4.227	UDP	62	Source port: 3671  Destination port: 59102
450	25.479014000	192.168.4.227	192.168.3.117	UDP	58	Source port: 59103  Destination port: 33867
452	25.495590000	192.168.3.117	192.168.4.227	UDP	50	Source port: 33867  Destination port: 59102
457	26.552969000	192.168.4.227	192.168.3.117	UDP	58	Source port: 51745  Destination port: 3671
```

```
Nov 26 18:38:37 knx knxd[26230]: Layer 0(011D6030,5839D6AD) Recv(026): 06 10 02 05 00 1A 08 01 C0 A8 04 E3 E6 DE 08 01 C0 A8 04 E3 E6 DF 04 04 02 00
Nov 26 18:38:37 knx knxd[26230]: Layer 1(011D6030,5839D6AD) Recv(020): 08 01 C0 A8 04 E3 E6 DE 08 01 C0 A8 04 E3 E6 DF 04 04 02 00
Nov 26 18:38:37 knx knxd[26230]: Layer 3(011B4FE0,5839D6AD) Allocate 7.0.101
Nov 26 18:38:37 knx knxd[26230]: Layer 8(011C5720,5839D6AD) Tunnel CONNECTION_REQ with 7.0.101
Nov 26 18:38:37 knx knxd[26230]: Layer 3(011B4FE0,5839D6AD) registerLayer2 0122BB58
Nov 26 18:38:37 knx knxd[26230]: Layer 3(011B4FE0,5839D6AD) registerLayer2 0122BB58 = 1
Nov 26 18:38:37 knx knxd[26230]: Layer 1(011D6030,5839D6AD) Send(014): 01 00 08 01 C0 A8 03 75 84 4B 04 04 70 65
Nov 26 18:38:37 knx knxd[26230]: Layer 0(011D6030,5839D6AD) Send(020): 06 10 02 06 00 14 01 00 08 01 C0 A8 03 75 84 4B 04 04 70 65
Nov 26 18:38:37 knx knxd[26230]: Layer 0(011C5890,5839D6AD) Recv(016): 06 10 02 07 00 10 01 00 08 01 C0 A8 04 E3 E6 DE
Nov 26 18:38:37 knx knxd[26230]: Layer 1(011C5890,5839D6AD) Recv(010): 01 00 08 01 C0 A8 04 E3 E6 DE
Nov 26 18:38:37 knx knxd[26230]: Layer 8(011C5720,5839D6AD) Invalid control address
Nov 26 18:38:37 knx knxd[26230]: Layer 1(011C5890,5839D6AD) Send(002): 01 15
Nov 26 18:38:37 knx knxd[26230]: Layer 0(011C5890,5839D6AD) Send(008): 06 10 02 08 00 08 01 15
Nov 26 18:38:38 knx knxd[26230]: Layer 0(011D6030,5839D6AE) Recv(016): 06 10 02 07 00 10 01 00 08 01 C0 A8 04 E3 CA 21
Nov 26 18:38:38 knx knxd[26230]: Layer 1(011D6030,5839D6AE) Recv(010): 01 00 08 01 C0 A8 04 E3 CA 21
Nov 26 18:38:38 knx knxd[26230]: Layer 8(011C5720,5839D6AE) Invalid control address
Nov 26 18:38:38 knx knxd[26230]: Layer 1(011D6030,5839D6AE) Send(002): 01 15
Nov 26 18:38:38 knx knxd[26230]: Layer 0(011D6030,5839D6AE) Send(008): 06 10 02 08 00 08 01 15
Nov 26 18:38:38 knx knxd[26230]: Layer 0(011D6030,5839D6AE) Recv(016): 06 10 02 09 00 10 01 00 08 01 C0 A8 04 E3 CA 21
Nov 26 18:38:38 knx knxd[26230]: Layer 1(011D6030,5839D6AE) Recv(010): 01 00 08 01 C0 A8 04 E3 CA 21
Nov 26 18:38:38 knx knxd[26230]: Layer 8(011C5720,5839D6AE) Invalid control address

```

**Comment by elupus** [comment]
This is with ETS 5.5.2. I wonder if it's worth checking the source port of the packet at all. Seems a bit fragile. I suspect ETS is using multiple sockets internally and keeps on re-allocating them.

**Comment by elupus** [comment]
It doesn't work after dropping the checks. It connects after a LONG time, but the busmonitor doesn't give any data. I'm starting to think this is some type of bug in ETS5.  Seems to work fine in NAT mode.

**Comment by elupus** [comment]
Okey.. Getting somewhat closer. ETS5 does not seem to like the port switch knxd is doing. It's issuing a connect_request against the standard port 3671, where it get's a response. But that response is indicating a new data port. Oddly enough, it seems to follow that data port and issue the connection_state request against that data port instead of the original control channel.

It then goes on to be utterly confused when packets arrive as responses to requests on the data channel. 

If you switch the target port in ETS5 to the dataport that knxd have opened in parallel to the control port (ie, no switch occurs on the connect), then all works well.

**Comment by elupus** [comment]
Since the discover socket also accepts any data. A simple hack is to just re-assign the EIBnetServer::Port variable after the discovery setup as such:

```
diff --git a/src/libserver/eibnetserver.cpp b/src/libserver/eibnetserver.cpp
index 429ea56..565b499 100644
--- a/src/libserver/eibnetserver.cpp
+++ b/src/libserver/eibnetserver.cpp
@@ -183,6 +183,7 @@ EIBnetServer::init (Layer3 *l3,
   if (!mcast->init ())
     goto err_out3;
 
+  Port = htons(port);
   this->tunnel = tunnel;
   this->route = route;
   this->discover = discover;
```

I think best would be to completely get rid of that extra socket for EIBnetServer. Current implementation doesn't support multiple control points anyway. So they might as well just use the same socket.

**Comment by smurfix** [comment]
On 27.11.2016 00:18, Joakim Plate wrote:
>
> I think best would be to completely get rid of that extra socket for
> EIBnetServer. Current implementation doesn't support multiple control
> points anyway. So they might as well just use the same socket.
>
I need a second port to send data packets from. Otherwise I won't be
able to run more than one knxd (or some other knx multicast user) on the
same host, (a) because the port numbers clash and (b) knxd needs to
filter its own transmissions. That's not possible if my data packets
always originate from port 3671.

So the basic idea is to bind one socket to port 3671, which will receive
multicast packets. All other communication shall use the second socket.

-- 
-- Matthias Urlichs

**Comment by GeminiServer** [comment]
Same issue here... would be cool if you could fix it

**Comment by elupus** [comment]
I wonder what would happen if we originate the connect reply from the
server socket instead of the multicast one directly. Firewalls might
misslike it. But that's what nat mode is for.

**Comment by elupus** [comment]
hmm. nope. it will not accept reply's on a different port for the connect.

**Comment by elupus** [comment]
Okey.. a summary 

* The spec doesn't really say anything about the case when explicitly configuring a knxd router instead of finding it using discovery. ETS5 doesn't know any other control point address then the one set in the GUI ie (3671 with default setup)

* knxd will have a random control endpoint? / data endpoint.
* ETS5 will invalidly target the announced data endpoint in the connect response for it's connection state requests instead of the (unknown) control endpoint
* ETS5 will originate it's connection state request from it's data endpoint (weird....)
* ETS5 will send it's control endpoint as the reply address in the connection state request
* ETS5 will be listening on the originating socket (the data endpoint) instead of the indicated control endpoint leading to a 10sec timeout with lack of response. 

i think we should put this issue on hold until next ETS5 is out, to see if they resolve the bugs or not. It would be good if we had a way to add a bug report for ETS5, but i've not found it yet.

**Comment by elupus** [comment]
I've opened a support issue on knx.org number #52200

**Comment by smurfix** [comment]
While we're waiting on that: anybody who is affected by this bug should try using ETS5 in multicast mode: attach to the local network interface, not a specific knxd instance.

**Comment by smurfix** [comment]
I'm also considering reverting the dual-port change, or possibly allow switching between the old and new version with an option. Does anybody actually need that (i.e. you need ETS5 but cannot use multicast)?

**Comment by elupus** [comment]
I cannot use multicast... Think VPN. That is different from the need of
multiple ports though.

Unicast tunnelling works fine against one port (at least against one
running binary).

That said, I don't see anything particularly wrong with the multiport
approach per say.

Current status from ets support.
"This "diagonal" request is done on purpose to ensure that the Windows
Firewall is opened on both the Control and the status port for incoming
Frames.

The timeout issue has to be investigated."

So they will likely resolve that particular issue via an update.





On Dec 12, 2016 00:18, "Matthias Urlichs" <notifications@github.com> wrote:

I'm also considering reverting the dual-port change, or possibly allow
switching between the old and new version with an option. Does anybody
actually need that (i.e. you need ETS5 but cannot use multicast)?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/knxd/knxd/issues/167#issuecomment-266316146>, or mute
the thread

**Comment by smurfix** [comment]
There's now single-port and multi-port options. Single-port, the old+compatible method, is the default until this is fixed. However I also found a possible alternate reason for the timeout, so if you could re-test v0.12.current with either of these flags I'd be grateful.

**Comment by elupus** [comment]
ETS 5.5.1 - Test Button
* Single port mode works (NAT and NON-NAT).
* Multi port mode does NOT work even thou it says OK (takes forever to complete)
* Multi port mode does work with NAT

There is a new ETS 5.5.2 release that has some fixes in regards to tunneling (likely due to my issue report). I'll update once i have that installed.


Other not working stuff not fully investigated
1. Device Info/Programming does not work in either modes (-B single did not solve it).
2. knxd does not send tunnel shutdown to remote tunnel server (likely error message "EIBnet: didn't de-register") causing tunnel connections to be depleated.
